File: examples/C/mtserver.c
//
//  Multithreaded Hello World server
//
#include "zhelpers.h"
#include <pthread.h>

static void *
worker_routine (void *context) {
    //  Socket to talk to dispatcher
    void *receiver = zmq_socket (context, ZMQ_REP);
    zmq_connect (receiver, "inproc://workers");

    while (1) {
        char *string = s_recv (receiver);
        printf ("Received request: [%s]\n", string);
        free (string);
        //  Do some 'work'
        sleep (1);
        //  Send reply back to client
        s_send (receiver, "World");
    }
    zmq_close (receiver);
    return NULL;
}

int main (void)
{
    void *context = zmq_init (1);

    //  Socket to talk to clients
    void *clients = zmq_socket (context, ZMQ_ROUTER);
    zmq_bind (clients, "tcp://*:5555");

    //  Socket to talk to workers
    void *workers = zmq_socket (context, ZMQ_DEALER);
    zmq_bind (workers, "inproc://workers");

    //  Launch pool of worker threads
    int thread_nbr;
    for (thread_nbr = 0; thread_nbr < 5; thread_nbr++) {
        pthread_t worker;
        pthread_create (&worker, NULL, worker_routine, context);
    }
    //  Connect work threads to client threads via a queue
    zmq_device (ZMQ_QUEUE, clients, workers);

    //  We never get here but clean up anyhow
    zmq_close (clients);
    zmq_close (workers);
    zmq_term (context);
    return 0;
}

File: examples/Perl/mtserver.pl
# Multithreaded Hello World server in Perl

use strict;
use warnings;
use v5.10;

use ZMQ::FFI;
use ZMQ::FFI::Constants qw(ZMQ_REP ZMQ_ROUTER ZMQ_DEALER);

use threads;

sub worker_routine {
    my ($context) = @_;

    # Socket to talk to dispatcher
    my $receiver = $context->socket(ZMQ_REP);
    $receiver->connect('inproc://workers');

    while (1) {
        my $string = $receiver->recv();

        say "Received request: [$string]";

        # Do some 'work'
        sleep 1;

        # Send reply back to client
        $receiver->send('World');
    }
}

my $context = ZMQ::FFI->new();

# Socket to talk to clients
my $clients = $context->socket(ZMQ_ROUTER);
$clients->bind('tcp://*:5555');

# Socket to talk to workers
my $workers = $context->socket(ZMQ_DEALER);
$workers->bind('inproc://workers');

# Launch pool of worker threads
for (1..5) {
    threads->create('worker_routine', $context);
}

# Connect work threads to client threads via a queue proxy
$context->proxy($clients, $workers);

# We never get here
